File: persistence/src/main/java/io/sparkled/persistence/sequence/impl/query/GetSongAudioBySequenceIdQuery.java
package io.sparkled.persistence.sequence.impl.query;

import io.sparkled.model.entity.SongAudio;
import io.sparkled.persistence.PersistenceQuery;
import io.sparkled.persistence.QueryFactory;

import java.util.Optional;

public class GetSongAudioBySequenceIdQuery implements PersistenceQuery<Optional<SongAudio>> {

    private final int sequenceId;

    public GetSongAudioBySequenceIdQuery(int sequenceId) {
        this.sequenceId = sequenceId;
    }

    @Override
    public Optional<SongAudio> perform(QueryFactory queryFactory) {
        SongAudio songAudio = queryFactory
                .select(qSongAudio)
                .from(qSequence)
                .innerJoin(qSongAudio).on(qSequence.songId.eq(qSongAudio.songId))
                .where(qSequence.id.eq(sequenceId))
                .fetchFirst();

        return Optional.ofNullable(songAudio);
    }
}

File: persistence/src/main/kotlin/io/sparkled/persistence/sequence/impl/query/GetSongAudioBySequenceIdQuery.kt
package io.sparkled.persistence.sequence.impl.query

import io.sparkled.model.entity.QSequence.Companion.sequence
import io.sparkled.model.entity.QSongAudio.Companion.songAudio
import io.sparkled.model.entity.SongAudio
import io.sparkled.persistence.PersistenceQuery
import io.sparkled.persistence.QueryFactory

class GetSongAudioBySequenceIdQuery(private val sequenceId: Int) : PersistenceQuery<SongAudio?> {

    override fun perform(queryFactory: QueryFactory): SongAudio? {
        return queryFactory
            .select<SongAudio?>(songAudio)
            .from(sequence)
            .innerJoin<SongAudio?>(songAudio).on(sequence.songId.eq(songAudio.songId))
            .where(sequence.id.eq(sequenceId))
            .fetchFirst()
    }
}
